sdk/eventhubs/azure_messaging_eventhubs/src/event_processor/load_balancer.rs

```diff
@@ -0,0 +1,1260 @@
+    duration: Duration,
+    consumer_client_details: ConsumerClientDetails,
+
+    // Random number generator used for load balancing - this is a test hook to allow for deterministic test results.
+    // In production, this should be a random number generator that is seeded with entropy.
+    // The mutex ensures that we can access a mutable RNG even if self is immutable.
+    rng: Mutex<Box<dyn RngCore + Send + Sync>>,
+}
+
+impl LoadBalancer {
+    /// Creates a new LoadBalancer instance.
+    pub fn new(
+        checkpoint_store: Arc<dyn CheckpointStore>,
+        consumer_client_details: ConsumerClientDetails,
+        processor_strategy: ProcessorStrategy,
+        duration: Duration,
+        rng: Option<Box<dyn RngCore + Send + Sync>>,
+    ) -> Self {
+        LoadBalancer {
+            checkpoint_store,
+            processor_strategy,
+            duration,
+            consumer_client_details,
+            rng: Mutex::new(rng.unwrap_or_else(|| Box::new(StdRng::from_entropy()))),
+        }
+    }
+
+    fn rng(&self) -> Result<MutexGuard<Box<dyn RngCore + Send + Sync>>> {
+        self.rng
+            .lock()
+            .map_err(|_| Error::message(AzureErrorKind::Other, "Failed to lock RNG mutex"))
+    }
+
+    async fn get_available_partitions(&self, partition_ids: &[&str]) -> Result<LoadBalancerInfo> {
+        trace!(
+            "[{}]Get available partitions for {}/{}/{}",
+            self.consumer_client_details.client_id,
+            self.consumer_client_details.fully_qualified_namespace,
+            self.consumer_client_details.eventhub_name,
+            self.consumer_client_details.consumer_group
+        );
+        let ownerships = self
+            .checkpoint_store
+            .list_ownerships(
+                self.consumer_client_details
+                    .fully_qualified_namespace
+                    .as_str(),
+                self.consumer_client_details.eventhub_name.as_str(),
+                self.consumer_client_details.consumer_group.as_str(),
+            )
+            .await?;
+
+        trace!("Found {} Ownerships", ownerships.len());
+
+        let mut unowned_or_expired: Vec<Ownership> = Vec::new();
+        let mut partitions_already_added: HashSet<String> = HashSet::new();
+        let mut grouped_by_owner: HashMap<String, Vec<Ownership>> = HashMap::new();
+        grouped_by_owner.insert(self.consumer_client_details.client_id.clone(), Vec::new());
+
+        for ownership in ownerships.iter() {
+            partitions_already_added.insert(ownership.partition_id.clone());
+
+            if let Some(last_modified_time) = ownership.last_modified_time {
+                if last_modified_time + self.duration < SystemTime::now() {
+                    unowned_or_expired.push(ownership.clone());
+                    continue;
+                }
+            }
+            // If the owner ID is empty, it means the partition is unowned.
+            if ownership.owner_id.is_empty() {
```

Comment:
Why isn't this an `Option<String>` then? That's idiomatic.

Reply:
Tip: FWIW, you can easily get a value or return in these cases using the `let` pattern instead of `if let` e.g.,

```rust
let Some(owner_id) = &ownership.owner_id else {
    unowned_or_expired.push(ownership.clone());
    continue; // at least, this might work; normally you have to `return` but maybe just exiting the parent scope is enough?
};

// now use `owner_id` as a `&str` like normal.
```

Reply:
That's a go-ism that got accidentally carried over.

Reply:
Oh, and the "let" syntax with an `else` clause is not one I've hit before, that's quite clever.